File: javascript/saddle-points/saddle-points.js
export const saddlePoints = (matrix) => {
  let saddlePoints = [];
  for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix[0].length; j++) {
      if (smallestInColumn(matrix, i, j) && largestInRow(matrix, i, j)) {
        saddlePoints.push({row: i + 1, column: j + 1})
      };
    };
  };
  return saddlePoints;
};

const smallestInColumn = (matrix, row, column) => {
  return matrix.map(matrixRow => matrixRow[column]).every(value => value >= matrix[row][column]);
}

const largestInRow = (matrix, row, column) => {
  return matrix[row].every(value => value <= matrix[row][column]);
}

File: python/saddle-points/saddle_points.py
def saddle_points(matrix):
    points = []
    for row_index, row in enumerate(matrix):
        if len(row) != len(matrix[0]):
            raise ValueError("irregular matrix")
        for column_index, column in enumerate(row):
            if is_tallest_in_row(row_index, column_index, matrix) and is_smallest_in_column(row_index, column_index, matrix):
                points.append({"row": row_index + 1, "column": column_index + 1})
    return points
            
def is_tallest_in_row(row_index, column_index, matrix):
    for element in matrix[row_index]:
        if element > matrix[row_index][column_index]:
            return False
    return True


def is_smallest_in_column(row_index, column_index, matrix):
    for row in matrix:
        if row[column_index] < matrix[row_index][column_index]:
            return False
    return True
    